packages/flutter_driver/lib/src/extension/extension.dart

```diff
@@ -369,6 +371,21 @@ class FlutterDriverExtension {
     return null;
   }
 
+  /// Returns a future that waits until frame is synced.
+  ///
+  /// Specifically, it checks:
+  /// * Whether the count of transient callbacks is zero.
+  /// * Whether there's no pending request for scheduling a new frame.
+  ///
+  /// We consider the frame is synced when both conditions listed above are met.
+  Future<Result> _waitUntilFrameSync(Command command) async {
+    await _waitUntilFrame(() {
+      return SchedulerBinding.instance.transientCallbackCount == 0
+          && !SchedulerBinding.instance.hasScheduledFrame;
```

Comment:
This will never return if there are ongoing animations, correct?  E.g. there's an indeterminate progress indicator, or blinking cursors are being allowed in the test for some reason.

Reply:
Another concern here is that by the time we get a response, this might no longer be true - I'm not sure if there's any way around that, but it seems like it would be a source of flakiness.

Reply:
Thanks for the review, Dan!

To your first question, I believe this method will never return if there are ongoing animations, but there's a [_frameSync](https://github.com/flutter/flutter/blob/c60063b72ded6215c0e7cde5f68f57d1972ef389/packages/flutter_driver/lib/src/extension/extension.dart#L167) flag that users can turn this off. And I also believe the current frame sync check [here](https://github.com/flutter/flutter/blob/c60063b72ded6215c0e7cde5f68f57d1972ef389/packages/flutter_driver/lib/src/extension/extension.dart#L239) probably should include this hasScheduledFrame check as well, but it seems not working well with the widgetTester's pump implementation today. On the other hand, in today's Espresso tests for Android native apps,  we disable animation by default and if users turn animation on, Espresso reports the app never goes idle. So it somewhat aligns with users' expectation, unless Flutter apps use animation heavily.

To your second question, Espresso talks to the driver extension API cross-process, so it's hard to keep truly synchronized, but I would think adding this API doesn't make things worse though :). If we could have this check closer to when we looking for an element [here](https://github.com/flutter/flutter/blob/c60063b72ded6215c0e7cde5f68f57d1972ef389/packages/flutter_driver/lib/src/extension/extension.dart#L239), probably would help reduce flakiness.

Reply:
Thanks for the explanation.  Maybe we could have a comment around this explaining some of that?

Reply:
Done. PTAL.

Reply:
```
If interacting
  /// with the app during animation is desirable, do NOT call this method. At
  /// the same time, you might also want to call the 'set_frame_sync' method to
  /// disable the frame sync while looking for UI elements.`
```
It might be better to replace this with the something like the prose from `runUnsynchronized` in driver.dart.  We could consider making that a template, or just borrowing a bit from it that makes sense here.  We also probably want to have a `See also` that links to `[FlutterDriver.runUnsynchronized]`.

Something like:

```
  /// This method relies on a Flutter Driver mechanism is called "frame sync", which
  /// waits for transient animations to finish. Persistent animations will cause this to
  /// wait forever.
  ///
  /// If a test needs to interact with the app, it should avoid this method and instead
  /// disable the frame sync using `set_frame_sync` method. See 
  /// [FlutterDriver.runUnsynchronized] for more details on how to do this.
  ///
  /// Disabling frame sync will require the test author to use some other method to
  /// avoid flakiness.
```

Reply:
Thanks! Reads way better :). PTAL.